File: 2024/js/day12/solution.js
/** @param {string} ornaments
 * @return {number} - The price of the tree
 */
function calculatePrice(ornaments) {
  const prices = { "*": 1, o: 5, "^": 10, "#": 50, "@": 100 };
  let ornaments_prices = [];
  for (let char of ornaments) {
    if (!Object.keys(prices).includes(char)) {
      return undefined;
    }
    ornaments_prices.push(prices[char]);
  }
  for (let i = 0; i < ornaments_prices.length; i++) {
    let current_price = ornaments_prices[i];
    let next_price = ornaments_prices[i + 1];
    if (current_price < next_price) {
      ornaments_prices[i] = current_price * -1;
    }
  }
  return ornaments_prices.reduce(
    (accumulator, currentValue) => accumulator + currentValue,
    0,
  );
}

File: 2024/python/day12/solution.py
def calculate_price(ornaments: str) -> int:
    prices = {"*": 1, "o": 5, "^": 10, "#": 50, "@": 100}
    ornaments_prices = []
    for char in ornaments:
        if char not in prices:
            return None
        ornaments_prices.append(prices[char])
    i = 0
    for price in ornaments_prices:
        if i + 1 < len(ornaments_prices):
            if price < ornaments_prices[i + 1]:
                ornaments_prices[i] = price * -1
        i += 1
    return sum(ornaments_prices)
